Add rewriteArraySort for Spark

boneanxs opened the pull request:
Support Spark `array_sort` to allow `lambda` function to sort elements.

Since Spark has different comparisons implementation than presto, we can't directly reuse presto `array_sort` logic to rewrite `lambda` function to a simple comparator if possible.

This pr tries to:
1. Move Presto `array_sort` to `velox/functions/lib` whereas both presto and spark can use it
2. Add a new option `nullsFirst` to support nulls to be placed at the start of the array(to support spark function `sort_array`
3. Extract the common logic of `SimpleComparisonMatcher` and move it to `velox/functions/lib`, and create different `SimpleComparisonChecker` for spark and presto to do the comparison match(e.g, `=` is `eq` in presto, but `equalto` in spark)
4. Add tests to cover spark rewrite function logic

boneanxs commented on velox/functions/sparksql/ArraySort.cpp:
```diff
@@ -278,6 +114,62 @@ std::shared_ptr<exec::VectorFunction> makeSortArray(
+    return rewritten;
+  }
+
+  VELOX_USER_FAIL(kNotSupported, lambda->toString())
```
Do we need to throw error out if the rewrite is not possible for spark?

boneanxs commented:
@PHILO-HE @rui-mo Hey, could you pls review this? We need this refactor to support spark `array_sort` function

rui-mo reviewed (commented):
>Since Spark has different comparisons implementation than presto

Could we add a block the PR description to describe the semantic difference?

rui-mo commented on velox/functions/lib/ArraySort.h:
```diff
@@ -0,0 +1,58 @@
+std::vector<std::shared_ptr<exec::FunctionSignature>> arraySortSignatures(
+    bool withComparator);
+
+} // namespace facebook::velox::functions
```
nit: add one empty line at the end.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -13,13 +13,17 @@
 #include "velox/vector/ConstantVector.h"
 
-namespace facebook::velox::functions::prestosql {
+namespace facebook::velox::functions::lib {
```
Shall we use the same namespace `facebook::velox::functions` as other files under functions/lib?

rui-mo commented on velox/functions/sparksql/tests/SortArrayTest.cpp:
```diff
@@ -153,7 +153,8 @@ TEST_F(SortArrayTest, array) {
 
-TEST_F(SortArrayTest, map) {
+// Map is unordered, so sorting is not supported.
+TEST_F(SortArrayTest, DISABLED_map) {
```
Does Spark throw for map type? If so, perhaps we could verify the exception in unit test.

boneanxs replied:
Yeap, spark also only allows orderable types to sort, while map type is not orderable inside spark.

```scala
checkArgumentDataTypes() match {
      case TypeCheckResult.TypeCheckSuccess =>
        argument.dataType match {
          case ArrayType(dt, _) if RowOrdering.isOrderable(dt) =>
//...
```

```sql
spark-sql> select sort_array(array(map(1,2)));
Error in query: cannot resolve 'sort_array(array(map(1, 2)), true)' due to data type mismatch: sort_array does not support sorting array of type map<int,int> which is not orderable; line 1 pos 7;
spark-sql> select array_sort(array(map(1,2)));
Error in query: cannot resolve '(namedlambdavariable() < namedlambdavariable())' due to data type mismatch: LessThan does not support ordering on type map<int,int>; line 1 pos 7;
```

Added the exception check in the unit tests

rui-mo reviewed (commented):
Thanks!

boneanxs commented:
> > Since Spark has different comparisons implementation than presto
> 
> Could we add a block the PR description to describe the semantic difference?

@rui-mo done, address the comments, pls have a look again.

rui-mo commented:
> Since Spark has different comparisons implementation than presto(see https://github.com/facebookincubator/velox/pull/5569)

@boneanxs I wonder if we are tackling the difference of NaN semantics in this PR. There is a plan in Velox to adjust its semantics, and some PRs have been merged. Perhaps we can fix the Presto function directly, seeing https://github.com/facebookincubator/velox/pull/7237.

boneanxs commented:
> > Since Spark has different comparisons implementation than presto(see #5569)
> 
> @boneanxs I wonder if we are tackling the difference of NaN semantics in this PR. There is a plan in Velox to adjust its semantics, and some PRs have been merged. Perhaps we can fix the Presto function directly, seeing #7237.

@rui-mo We might still need a special spark rewrite arraySort logic even the NaN semantics difference is fixed, given:
1. We have different comparison implementation now, gluten maps spark's implementation instead of presto's, it could throw errors since presto's `rewriteArraySortCall` can't correctly identify spark's comparison names.

e.g. for the expression `array_sort(array(), (left, right) -> if (left > right, 1, if(left < right, -1, 0)))`, gluten could transform the expression to `array_sort(array(), lambda ROW<left:INTEGER,right:INTEGER> -> if(greaterthan(left, right), 1, if(lessthan(left, right), -1, 0)))`, while presto side can only identify `array_sort(array(), lambda ROW<left:INTEGER,right:INTEGER> -> if(gt(left, right), 1, if(lt(left, right), -1, 0)))`(The function names are different now).

Do we have plan to fix all spark comparison functions after the NaN semantics is unified? And I'm thinking it might be a long term to address it?

3. Is it possible spark could have other funtions that is different from presto and we need to specially handle it in `rewriteArraySort`, given many functions have different meaning and some functions independently inside spark, it's possible we can do some optimization inside spark while presto doesn't support or vice visa. So separating the `rewriteArraySort` can give us this possibility.

rui-mo commented on velox/functions/sparksql/SimpleComparisonChecker.h:
```diff
@@ -0,0 +1,113 @@
+  std::string* op_;
+};
+
+class SparkSimpleComparisonChecker : public SimpleComparisonChecker {
```
It seems the only difference between SparkSimpleComparisonChecker and PrestoSimpleComparisonChecker is the names. I wonder if we can define some pure virtual functions in the base class to return function names and override them for Spark and Presto in the derived classes. In this way, the other parts of code can be reused.

rui-mo commented on velox/functions/sparksql/SimpleComparisonChecker.h:
```diff
@@ -0,0 +1,113 @@
+namespace facebook::velox::functions::sparksql {
+using namespace facebook::velox::functions;
+
+class SparkComparisonMatcher : public Matcher {
```
Same for SparkComparisonMatcher and ComparisonMatcher.

rui-mo commented on velox/functions/sparksql/ArraySort.cpp:
```diff
@@ -278,6 +114,62 @@ std::shared_ptr<exec::VectorFunction> makeSortArray(
+      "array_sort with comparator lambda that cannot be rewritten "
+      "into a transform is not supported: {}";
+
+  auto checker = std::make_unique<SparkSimpleComparisonChecker>();
```
I wonder if we can add a parameter of MatcherPtr for the rewriteArraySortCall function to provide custom matcher for Presto and Spark. In this way, the other parts of code can also be reused.

rui-mo reviewed (commented):
Thanks.

rui-mo commented on velox/functions/sparksql/SimpleComparisonMatcher.h:
```diff
@@ -0,0 +1,64 @@
+  }
+
+ public:
+  ~SparkSimpleComparisonChecker() override = default;
```
nit: move it to the start of class for readability.

rui-mo reviewed (commented):
Thanks. Added several comments.

boneanxs commented on velox/functions/lib/ArraySort.cpp:
```diff
@@ -171,7 +190,7 @@ void applyScalarType(
       } else {
         bits::fillBits(rawBits, startRow, startRow + numOneBits, true);
-        bits::fillBits(rawBits, endZeroRow, endRow, false);
+        bits::fillBits(rawBits, startRow + numOneBits, endRow, false);
```
Yeap, this is required here now since spark `sort_array` could pass ascending, and it covered in `SortArrayTest`

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -112,6 +112,26 @@ Array Functions
         SELECT array_sort(array(NULL, 1, NULL)); -- [1, NULL, NULL]
         SELECT array_sort(array(NULL, 2, 1)); -- [1, 2, NULL]
 
+.. spark:function:: array_sort(array(T), function(T,U)) -> array(T)
```
Use `array_sort(array(T)[, function(T,U)])` to illustrate function is optional.

rui-mo replied:
It makes sense to me to separate them. Thanks.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -112,6 +112,26 @@ Array Functions
+    :noindex:
+
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. U must be an orderable type. Null elements will be placed at the end of
```
Would you clarify what kind of lambda is expected here?

nit: U -> \``U\`` to referring to the variable

Would you document the NaN behavior?

Also document the behavior if the comparator function returns null.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -112,6 +112,26 @@ Array Functions
+    order. U must be an orderable type. Null elements will be placed at the end of
+    the returned array. ::
+
+        SELECT array_sort(array('cat', 'leopard', 'mouse'), x -> length(x)); -- ['cat', 'mouse', 'leopard']
```
Perhaps add examples for NaN and null values.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -112,6 +112,26 @@ Array Functions
+    :noindex:
+    
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. U must be an orderable type. Null elements will be placed at the end of
```
> in ascending order

Is it true only when function is not provided?

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -112,6 +112,26 @@ Array Functions
+    order. U must be an orderable type. Null elements will be placed at the end of
+    the returned array.
+    The function attempts to analyze the lambda function and rewrite it into a simpler call that 
+    specifies the sort-by expression(like array_sort(array(T), function(T,U)) -> array(T) mentioned above). ::
```
> The function attempts to analyze the lambda function and rewrite it into a simpler call that 
    specifies the sort-by expression(like array_sort(array(T), function(T,U)) -> array(T) mentioned above)

rui-mo replied:
> like array_sort(array(T), function(T,U)) -> array(T) mentioned above

Thanks for the explain. I feel this example is not clear enough, perhaps we can take a concrete example like `(left, right) -> if(length(left) > length(right), 1, if(length(left) < length(right), -1, 0))` will be rewritten as ...

rui-mo commented on velox/functions/lib/ArraySort.cpp:
```diff
@@ -221,8 +240,13 @@ class ArraySortFunction : public exec::VectorFunction {
+      bool throwOnNestedNull)
+      : ascending_{ascending},
+        nullsFirst_{nullsFirst},
+        throwOnNestedNull_(throwOnNestedNull) {}
```
nit: {} and () are used for different variables. Maybe change them as the same style.

rui-mo commented on velox/functions/lib/ArraySort.cpp:
```diff
@@ -495,9 +505,36 @@ core::CallTypedExprPtr asArraySortCall(
+    const core::QueryConfig& /*config*/,
+    bool ascending,
+    bool throwOnNestedNull) {
+  VELOX_CHECK_EQ(inputArgs.size(), 2);
```
nit: use VELOX_USER_CHECK_EQ instead as it does not indicate a bug but a user error.

rui-mo commented on velox/functions/lib/ArraySort.h:
```diff
@@ -0,0 +1,74 @@
+ * placed last.
+ * @param throwOnNestedNull If true, throw an exception if a nested null is
+ * encountered.
+ */
```
nit: use `///` for multiple-line comments of a public API. Ditto for the others.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -73,11 +77,19 @@ class ComparisonMatcher : public Matcher {
+  // Check if the given name is a comparison expression.
+  // Making this virtual to allow derived classes to override this since
+  // the name of the comparison expression can be different(e.g. in Spark `eq`
+  // is `equalTo`)
```
Comment suggestion:
>   // Checks if the given name specifies a comparison expression. Can be
  // overriden to use different function names for Spark.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -184,136 +198,91 @@ class ComparisonConstantMatcher : public Matcher {
   }
 
-  const auto eq = prefix + "eq";
+  /// Returns true for a < b -> -1.
```
nit: use '//' for single-line comment.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -184,136 +198,91 @@ class ComparisonConstantMatcher : public Matcher {
-  return {{transform, b1}};
-}
+ public:
+  virtual ~SimpleComparisonChecker() = default;
```
nit: add one empty line after it.

rui-mo commented on velox/functions/sparksql/Register.cpp:
```diff
@@ -327,10 +328,17 @@ void registerFunctions(const std::string& prefix) {
-      prefix + "array_sort", arraySortSignatures(), makeArraySort);
+      prefix + "array_sort", arraySortSignatures(true), makeArraySortAsc);
+  exec::registerStatefulVectorFunction(
+      prefix + "array_sort_desc", arraySortDescSignatures(), makeArraySortDesc);
```
Do we have corresponding function for `array_sort_desc` in Spark?

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -112,6 +112,26 @@ Array Functions
+
+        SELECT array_sort(array('cat', 'leopard', 'mouse'), x -> length(x)); -- ['cat', 'mouse', 'leopard']
+
+.. spark:function:: array_sort(array(T), function(T,T,U)) -> array(T)
```
I only find below signature is registered. Am I miss something?
> array_sort_desc(array(T), function(T,U))

boneanxs replied:
It's here:
https://github.com/facebookincubator/velox/blob/0093ee9348c705d460a204aef9e17e595aedfcce/velox/functions/sparksql/Register.cpp#L331

rui-mo reviewed (commented):
Thanks. Added several questions.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -104,13 +104,33 @@ Array Functions
+    :noindex:
+
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.  ::
```
Would you document the behavior for nested nulls? Perhaps also add examples & tests if not added.

Presto array_sort documentation: https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/presto/array.rst?plain=1#L175-L200

rui-mo commented on velox/functions/lib/ArraySort.cpp:
```diff
@@ -354,6 +380,7 @@ class ArraySortLambdaFunction : public exec::VectorFunction {
         *flatArray,
         *newElements,
         ascending_,
+        false,
```
nit: false \/\*nullsFirst\*\/

rui-mo commented on velox/functions/lib/ArraySort.cpp:
```diff
@@ -495,9 +505,36 @@ core::CallTypedExprPtr asArraySortCall(
+    bool ascending,
+    bool nullsFirst,
+    bool throwOnNestedNull) {
+  return create(inputArgs, ascending, nullsFirst, throwOnNestedNull);
```
Do we need this function as it just calls another function? Perhaps moves the logic of create into makeArraySort and remove create function.

rui-mo commented on velox/functions/lib/ArraySort.h:
```diff
@@ -0,0 +1,74 @@
+    bool nullsFirst,
+    bool throwOnNestedNull);
+
+/// Make array_sort with a lambda function.
```
nit: Make -> Creates

rui-mo commented on velox/functions/lib/ArraySort.h:
```diff
@@ -0,0 +1,74 @@
+/// into
+///     array_sort(a, x -> length(x))
+///
+/// Returns new expression or nullptr if rewrite is not possible.
```
Comment suggestion:
> Returns the rewritten expression. If rewrite is not possible, returns nullptr.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -13,13 +13,17 @@
+namespace facebook::velox::functions {
 
-namespace {
+struct SimpleComparison {
```
Perhaps add comments for these APIs in the header.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -73,11 +77,17 @@ class ComparisonMatcher : public Matcher {
+        name == prefix_ + "gt";
+  }
+
   bool match(const core::TypedExprPtr& expr) override {
```
Perhaps move the implementation to cpp file if it is not a small function to be inlined. Ditto for the others.

rui-mo commented on velox/functions/sparksql/ArraySort.cpp:
```diff
@@ -14,243 +14,64 @@
+      name,
+      inputArgs,
+      config,
+      /*ascending=*/true,
```
nit: true /\*ascending\*/

rui-mo commented on velox/functions/sparksql/ArraySort.cpp:
```diff
@@ -278,6 +99,13 @@ std::shared_ptr<exec::VectorFunction> makeSortArray(
+      config,
+      /*ascending=*/ascending,
+      /*nullsFirst=*/nullsFirst,
+      /*throwOnNestedNull=*/false);
```
Ditto.

rui-mo commented:
@boneanxs Would you like to update this PR? Thanks.

boneanxs commented:
> @boneanxs Would you like to update this PR? Thanks.

Oh, forgot it. Sure, will update it recently

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -125,13 +125,34 @@ Array Functions
+    :noindex:
+
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.  ::
```
NULL or NaN for floating type?

boneanxs replied:
Currently for lambda function returned values, Nan is not handled. Do we need to handle `NaN` since it shouldn't returned by lambda functions

boneanxs replied:
Oh, sorry, I overlooked this before, for array_sort with lambda functions, it supports sorting with `NaN` in `SimpleVector.comparePrimitiveAsc` and follows the logic of `NaN is before NULL`, I also add a test to cover this.

boneanxs replied:
Let me explain more here. After looking into the presto/spark implementation, they both say

> It returns -1, 0, or 1 as the first nullable element is less than, equal to, or greater than the second nullable element. If the comparator function returns other values (including NULL), the query will fail and raise an error

see [presto](https://prestodb.io/docs/current/functions/array.html#array_sort) and [spark](https://docs.databricks.com/en/sql/language-manual/functions/array_sort.html) (Though spark says it doesn't support returning null values, but it doesn't throw errors for query like `SELECT array_sort(ARRAY ('bc', 'ab', 'dc'), (x, y) -> IF(x < y, 1, IF(x = y, 0, null)))` in Spark3.2 which might be a bug)

So null values and NaN shouldn't be return for lambda function `function(T,T, int)`, and in `SimpleComparisonMatcher`, we do the match that the return value must be int.

`SimpleComparisonMatcher` could optimize `function(T,T, int)` to `function(T, U)` where `U` is orderable(not limited to int), it's possible that it creates float values, such as `function(float, float, int)`: `IF( x > y, 1, IF(x < y, -1, 0))` will be optimized to `function(float, float)`: `x -> x`, at such point, they should still be the same since both goes into `SimpleVector.compare` to do the comparison(except NULLs are filtered in `ArraySort.sortElements` in advance to respect `nullsFirst` flag). And inside `SimpleVector.compare`, `NaN` is smaller than `Null`.

rui-mo replied:
I tried `SELECT array_sort(ARRAY ('bc', 'ab', 'dc'), (x, y) -> IF(x < y, 1, IF(x = y, 0, null)))` in Spark 3.5 and got below exception. Would you like the add a unit test for this case to make sure exception is thrown?
> Caused by: org.apache.spark.SparkException: [COMPARATOR_RETURNS_NULL] The comparator has returned a NULL for a comparison between dc and dc. It should return a positive integer for "greater than", 0 for "equal" and a negative integer for "less than". To revert to deprecated behavior where NULL is treated as 0 (equal), you must set "spark.sql.legacy.allowNullComparisonResultInArraySort" to "true".

I also notice Spark requires the function must return integer type, would you like to confirm?
https://github.com/apache/spark/blob/branch-3.5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala#L412-L421

boneanxs replied:
Sorry to miss this comment, yes, for the integer type handling, Spark itself will throw the exception if the return type is not integer type(and it handles return type in analyze stage). Also, velox can't rewrite the lambda function if the lambda function doesn't match, I add a test in `ArraySortTest.unsupporteLambda` to ensure it.

As for null returned by comparator, it's fixed by https://github.com/apache/spark/pull/36812 since Spark 3.2.2( I use 3.2.1 so it can pass). Spark 3.5 doesn't have this issue.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -125,13 +125,34 @@ Array Functions
+    
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.
+    The function attempts to analyze the lambda function and rewrite it into a simpler call that 
```
rewrite -> rewrites

boneanxs replied:
Oh, I mean `attemps to analyze and attemps to rewrite`

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -125,13 +125,34 @@ Array Functions
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.
+    The function attempts to analyze the lambda function and rewrite it into a simpler call that 
+    specifies the sort-by expression(like array_sort(array(T), function(T,U)) -> array(T) mentioned above). for example, ``(left, right) -> if(length(left) > length(right), 1, if(length(left) < length(right), -1, 0))`` will be rewritten to ``x -> length(x)`` ::
```
nit: add white-space after 'expression'

rui-mo replied:
> like array_sort(array(T), function(T,U)) -> array(T) mentioned above

Perhaps change as below to refer to the last function.

like :spark:func:\`array_sort(array(T), function(T,U)) -> array(T)\`

rui-mo replied:
nit: add dot at the end for this sentence.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -125,13 +125,34 @@ Array Functions
+    The function attempts to analyze the lambda function and rewrite it into a simpler call that 
+    specifies the sort-by expression(like array_sort(array(T), function(T,U)) -> array(T) mentioned above). for example, ``(left, right) -> if(length(left) > length(right), 1, if(length(left) < length(right), -1, 0))`` will be rewritten to ``x -> length(x)`` ::
+
+        SELECT array_sort(array('cat', 'leopard', 'mouse'), (left, right) -> if(length(left) > length(right), 1, if(length(left) < length(right), -1, 0))); ['cat', 'mouse', 'leopard']
```
add '-- ' before the result of example.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -125,13 +125,34 @@ Array Functions
 
     Returns an array which has the sorted order of the input array(E). The elements of array(E) must
-    be orderable. Null elements will be placed at the end of the returned array. ::
+    be orderable. Null/NaN elements will be placed at the end of the returned array. ::
```
> Null/NaN elements will be placed at the end

For floating point data, which will be put at the end if null and NaN both exist?

boneanxs replied:
NaN is before NULL, there's a test already covers this, `floatingPointInput` in `ArraySortTestData`

rui-mo reviewed (commented):
Some comments on the documentation.

rui-mo commented on velox/functions/lib/ArraySort.h:
```diff
@@ -0,0 +1,74 @@
+    bool ascending,
+    bool throwOnNestedNull);
+
+/// Signatures for array_sort function.
```
nit: Returns signatures for ...

rui-mo commented on velox/functions/lib/ArraySort.h:
```diff
@@ -0,0 +1,74 @@
+/// into
+///     array_sort(a, x -> length(x))
+///
+/// Returns the rewritten expression. If rewrite is not possible, throw errors.
```
throw errors -> throws a user error

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -0,0 +1,201 @@
+
+using MatcherPtr = std::shared_ptr<Matcher>;
+
+/// Matches the if expression that children match the given matchers.
```
nit: use // for single-line comment.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -0,0 +1,201 @@
+
+using IfMatcherPtr = std::shared_ptr<IfMatcher>;
+
+/// Matches comparison expression that represents `eq`, `lt`, or `gt`.
```
Ditto.

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  // Checks if the given name specifies a comparison expression. Can be
+  // overriden to use different function names for Spark.
```
use ///

rui-mo commented on velox/functions/lib/SimpleComparisonMatcher.h:
```diff
@@ -0,0 +1,201 @@
+      std::vector<std::shared_ptr<Matcher>>& matchers);
+};
+
+using MatcherPtr = std::shared_ptr<Matcher>;
```
Please avoid `using ...` in the header file. Ditto for the others like this.

rui-mo commented on velox/functions/sparksql/tests/ArraySortTest.cpp:
```diff
@@ -104,10 +105,14 @@ TEST_F(ArraySortTest, array) {
 }
 
-TEST_F(ArraySortTest, map) {
+// Map is unordered, so sorting is not supported.
```
nit: unordered -> not orderable

rui-mo commented on velox/functions/sparksql/tests/SortArrayTest.cpp:
```diff
@@ -153,13 +154,16 @@ TEST_F(SortArrayTest, array) {
 }
 
-TEST_F(SortArrayTest, map) {
+// Map is unordered, so sorting is not supported.
```
ditto

rui-mo commented on velox/functions/sparksql/tests/ArraySortTest.cpp:
```diff
@@ -140,5 +145,66 @@ TEST_F(ArraySortTest, constant) {
+    SCOPED_TRACE(name);
+    SCOPED_TRACE(lambdaExpr);
+    auto result = evaluate(fmt::format("{}(c0, {})", name, lambdaExpr), data);
+    assertEqualVectors(sortedAsc, result);
```
It looks like the only difference between 'testAsc' and 'testDesc' is the expected result. Perhaps make it a parameter and reuse the test lambda.

rui-mo reviewed (commented):
Thanks for iterating. Some minors and the others look good!

boneanxs commented on velox/docs/functions/spark/array.rst:
```diff
@@ -125,13 +125,34 @@ Array Functions
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.  ::
+
+        SELECT array_sort(array('cat', 'leopard', 'mouse'), x -> length(x)); -- ['cat', 'mouse', 'leopard']
```
This function is only used by velox to rewrite `array_sort(array(T), function(T,T,U))` to `array_sort(array(T), function(T,U)) `, it's not supported by spark

rui-mo replied:
Perhaps clarify that in the documentation, and remove the unsupported query example.

rui-mo reviewed (commented):
@boneanxs Thanks for iterating.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,31 @@ Array Functions
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.
+    The function attempts to analyze the lambda function and rewrite it into a simpler call that 
+    specifies the sort-by expression (like :spark:func:`array_sort(array(T), function(T,U)) -> array(T)`). For example, ``(left, right) -> if(length(left) > length(right), 1, if(length(left) < length(right), -1, 0))`` will be rewritten to ``x -> length(x)``. If rewrite is not possible, a user error will be thrown ::
```
@boneanxs Thanks for your continuous work! Would you document the limitation of current rewriting on the null handling?

boneanxs replied:
done

boneanxs commented:
Hey @rui-mo , any more comments for this?

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,34 @@ Array Functions
 
     Returns an array which has the sorted order of the input array(E). The elements of array(E) must
-    be orderable. Null elements will be placed at the end of the returned array. ::
+    be orderable. Null/NaN elements will be placed at the end of the returned array, with NaN elements appearing before Null elements for float types. ::
```
nit: 
Null -> NULL
Null/NaN -> NULL and NaN
float types -> floating-point types

Ditto for the others.

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,34 @@ Array Functions
 
         SELECT array_sort(array(1, 2, 3)); -- [1, 2, 3]
         SELECT array_sort(array(3, 2, 1)); -- [1, 2, 3]
         SELECT array_sort(array(2, 1, NULL); -- [1, 2, NULL]
```
One `)` is missing in this example. Not your change, but could you help fix it? Thanks.

PHILO-HE reviewed (commented):
Just reviewed the added doc. Could you compile the rst file to see the added content is well displayed (including hyper link) in the generated doc? Thanks!

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,34 @@ Array Functions
+    :noindex:
+
+    Returns the array sorted by values computed using specified lambda in ascending order. ``U`` must be an orderable type.
+    Null/NaN elements returned by the lambda function will be placed at the end of the returned array, with NaN elements appearing before Null elements. This functions is not supported in Spark and is only used inside velox for rewring :spark:func: array_sort(array(T), function(T,T,U)) as :spark:func: array_sort(array(T), function(T,U)). ::
```
> for rewring :spark:func: array_sort(array(T), function(T,T,U)) as :spark:func: array_sort(array(T), function(T,U))

The two references do not work well. Perhaps refer the style for array_sort on L159.

rui-mo reviewed (commented):
Hi @boneanxs, I added some nits. And do we have feedback for https://github.com/facebookincubator/velox/pull/10138#discussion_r1857748931?

PHILO-HE commented:
@boneanxs, could you file a pr in Gluten to enable this function and also to see its CI feedback? Thus, we can make sure Spark UTs pass. In Gluten pr, your personal Velox branch with this patch should be referenced.

boneanxs commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,34 @@ Array Functions
+    :noindex:
+    
+    Returns the array sorted by values computed using specified lambda in ascending
+    order. ``U`` must be an orderable type. If the value from the lambda function is NULL, the element will be placed at the end.
```
Oh, not the same, it means the value returned from the rewritten function, such as `x -> length(x)` in the example.

I update here to make it more clear

boneanxs commented:
> Just reviewed the added doc. Could you compile the rst file to see the added content is well displayed (including hyper link) in the generated doc? Thanks!

Yes, it's displayed well

![Screenshot 2025-02-14 at 11 34 21](https://github.com/user-attachments/assets/1247dbaf-11ba-4a95-bc49-9427867a8c82)

> @boneanxs, could you file a pr in Gluten to enable this function and also to see its CI feedback? Thus, we can make sure Spark UTs pass. In Gluten pr, your personal Velox branch with this patch should be referenced.

There's a pr test it before, https://github.com/apache/incubator-gluten/pull/8526, and it all passes.

boneanxs commented:
> Hi @boneanxs, I added some nits. And do we have feedback for [#10138 (comment)](https://github.com/facebookincubator/velox/pull/10138#discussion_r1857748931)?

@rui-mo Oh, sorry I miss that before, added comment now. Can help review it again :)

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,36 @@ Array Functions
+
+    Returns the array sorted by values computed using specified lambda in ascending order. ``U`` must be an orderable type.
+    NULL and NaN elements returned by the lambda function will be placed at the end of the returned array, with NaN elements appearing before NULL elements.
+    This function is not supported in Spark and is only used inside Velox for rewring :spark:func:`array_sort(array(T), function(T,T,U)) -> array(T)` as :spark:func:`array_sort(array(T), function(T,U))`. ::
```
typo for rewring

boneanxs replied:
done

rui-mo commented on velox/docs/functions/spark/array.rst:
```diff
@@ -134,13 +134,36 @@ Array Functions
+
+    Returns the array sorted by values computed using specified lambda in ascending order. ``U`` must be an orderable type.
+    NULL and NaN elements returned by the lambda function will be placed at the end of the returned array, with NaN elements appearing before NULL elements.
+    This function is not supported in Spark and is only used inside Velox for rewring :spark:func:`array_sort(array(T), function(T,T,U)) -> array(T)` as :spark:func:`array_sort(array(T), function(T,U))`. ::
```
Is the second reference correct, and shall we also include its return type?
![Screenshot 2025-02-17 at 15 07 10](https://github.com/user-attachments/assets/b0b71089-18ee-4259-9dbd-48e77029a291)

boneanxs replied:
done

rui-mo reviewed (approved):
Thanks. Just two nits.

xiaodouchen commented on velox/functions/lib/SimpleComparisonMatcher.cpp:
```diff
@@ -0,0 +1,218 @@
+
+bool IfMatcher::match(const core::TypedExprPtr& expr) {
+  if (auto call = dynamic_cast<const core::CallTypedExpr*>(expr.get())) {
+    if (call->name() == "if" && allMatch(call->inputs(), inputMatchers_)) {
```
Can the "if" here also extract the name for separate comparisons, similar to other comparison operators?

boneanxs replied:
Sorry, I might not get your meaning, why we need to extract `if` here, since both `Presto` and `Spark` use the same `if` comparator

xiaodouchen replied:
In other systems, this expression might be a switch. By extracting this name, the Matcher could become more generic.

boneanxs replied:
If it's Switch, this could be a different Matcher instead of IfMatcher. Actually we should also support SwitchMatcher to rewrite the array_sort, but I'm considering to do it in a separate pr.

boneanxs commented:
@rui-mo Hey, all comments are addressed, can we continue to merge it? Thanks!

rui-mo commented:
@pedroerp @Yuhta Do you have any further comment? If not, we may mark it as read-to-merge. Thanks!

boneanxs commented:
Gentle ping @pedroerp @Yuhta

rui-mo commented:
@boneanxs Can you rebase this PR? Thanks.

boneanxs commented:
> @boneanxs Can you rebase this PR? Thanks.

done @rui-mo

rui-mo commented:
@Yuhta @pedroerp We can merge this PR if no further comment. Thanks.

kagamiori commented:
Hi @boneanxs, I got an internal test error saying that the comparator used with std::sort or std::lower_bound in sort_array() doesn't establish strict weak ordering. Could you take a look? Below is the error message and stack trace I got. Thanks!
```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from SortArrayTest
[ RUN      ] SortArrayTest.array

terminate called after throwing an instance of 'facebook::velox::VeloxUserError'
  what():  Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: strict weak ordering: ((__x LT __y) && (__y LT __x)) != false
Retriable: False
Context: Top-level Expression: sort_array(c0)
Function: 
File: UNKNOWN
Stack trace:
Stack trace has been disabled. Use --velox_exception_user_stacktrace_enabled=true to enable it.

*** Aborted at 1741823790 (Unix time, try 'date -d @1741823790') ***
*** Signal 6 (SIGABRT) (0x75590000263e) received by PID 9790 (pthread TID 0x7f5302ff3e80) (linux TID 9790) (maybe from PID 9790, UID 30041) (code: -6), stack trace: ***
    @ 000000000000f607 folly::symbolizer::(anonymous namespace)::innerSignalHandler(int, siginfo_t*, void*)
                       ./fbcode/folly/debugging/symbolizer/SignalHandler.cpp:453
    @ 000000000000dd81 folly::symbolizer::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)
                       ./fbcode/folly/debugging/symbolizer/SignalHandler.cpp:474
    @ 000000000004455f (unknown)
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/signal/../sysdeps/unix/sysv/linux/libc_sigaction.c:8
                       -> /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c
    @ 000000000009c993 __GI___pthread_kill
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/nptl/pthread_kill.c:46
    @ 00000000000444ac __GI_raise
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/signal/../sysdeps/posix/raise.c:26
    @ 000000000002c432 __GI_abort
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/stdlib/abort.c:79
    @ 00000000000a3fc4 __gnu_cxx::__verbose_terminate_handler()
                       /home/engshare/third-party2/libgcc/11.x/src/gcc-11.x/x86_64-facebook-linux/libstdc++-v3/libsupc++/../../.././libstdc++-v3/libsupc++/vterminate.cc:95
    @ 00000000000a1b29 __cxxabiv1::__terminate(void (*)())
                       /home/engshare/third-party2/libgcc/11.x/src/gcc-11.x/x86_64-facebook-linux/libstdc++-v3/libsupc++/../../.././libstdc++-v3/libsupc++/eh_terminate.cc:48
    @ 00000000000a1b94 std::terminate()
                       /home/engshare/third-party2/libgcc/11.x/src/gcc-11.x/x86_64-facebook-linux/libstdc++-v3/libsupc++/../../.././libstdc++-v3/libsupc++/eh_terminate.cc:58
    @ 00000000000a1eb1 __cxa_rethrow
                       /home/engshare/third-party2/libgcc/11.x/src/gcc-11.x/x86_64-facebook-linux/libstdc++-v3/libsupc++/../../.././libstdc++-v3/libsupc++/eh_throw.cc:133
    @ 0000000001586f72 facebook::velox::exec::Expr::applyFunction(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
                       ./fbcode/velox/expression/Expr.cpp:1529
    @ 0000000001598159 facebook::velox::exec::Expr::evalAllImpl(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
                       ./fbcode/velox/expression/Expr.cpp:1442
    @ 000000000158d46a facebook::velox::exec::Expr::evalAll(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
                       ./fbcode/velox/expression/Expr.cpp:1386
    @ 00000000015907b3 facebook::velox::exec::Expr::evalWithNulls(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
                       ./fbcode/velox/expression/Expr.cpp:1176
    @ 000000000158db0a facebook::velox::exec::Expr::evalEncodings(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&)
                       ./fbcode/velox/expression/Expr.cpp:1093
    @ 000000000158cca3 facebook::velox::exec::Expr::eval(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::shared_ptr<facebook::velox::BaseVector>&, facebook::velox::exec::ExprSet const*)
                       ./fbcode/velox/expression/Expr.cpp:834
    @ 0000000001550a20 facebook::velox::exec::ExprSet::eval(int, int, bool, facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > >&)
                       ./fbcode/velox/expression/Expr.cpp:1937
    @ 0000000000799021 facebook::velox::exec::ExprSet::eval(facebook::velox::SelectivityVector const&, facebook::velox::exec::EvalCtx&, std::vector<std::shared_ptr<facebook::velox::BaseVector>, std::allocator<std::shared_ptr<facebook::velox::BaseVector> > >&)
                       fbcode/velox/expression/Expr.h:719
    @ 0000000000797bd7 facebook::velox::functions::test::FunctionBaseTest::evaluate(facebook::velox::exec::ExprSet&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
                       fbcode/velox/functions/prestosql/tests/utils/FunctionBaseTest.h:335
    @ 0000000000797327 std::shared_ptr<facebook::velox::BaseVector> facebook::velox::functions::test::FunctionBaseTest::evaluateImpl<facebook::velox::exec::ExprSet>(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
                       fbcode/velox/functions/prestosql/tests/utils/FunctionBaseTest.h:418
    @ 0000000000796bb7 facebook::velox::functions::test::FunctionBaseTest::evaluate(std::shared_ptr<facebook::velox::core::ITypedExpr const> const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
                       fbcode/velox/functions/prestosql/tests/utils/FunctionBaseTest.h:96
    @ 0000000000795806 facebook::velox::functions::test::FunctionBaseTest::evaluate(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<facebook::velox::RowVector> const&, std::optional<facebook::velox::SelectivityVector> const&)
                       fbcode/velox/functions/prestosql/tests/utils/FunctionBaseTest.h:106
    @ 0000000001872308 facebook::velox::functions::sparksql::test::(anonymous namespace)::SortArrayTest::testSortArray(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<facebook::velox::BaseVector> const&, std::shared_ptr<facebook::velox::BaseVector> const&)
                       ./fbcode/velox/functions/sparksql/tests/SortArrayTest.cpp:54
    @ 00000000018714f3 facebook::velox::functions::sparksql::test::(anonymous namespace)::SortArrayTest::testSortArray(std::shared_ptr<facebook::velox::BaseVector> const&, std::shared_ptr<facebook::velox::BaseVector> const&, std::shared_ptr<facebook::velox::BaseVector> const&)
                       ./fbcode/velox/functions/sparksql/tests/SortArrayTest.cpp:40
    @ 0000000001899fbe facebook::velox::functions::sparksql::test::(anonymous namespace)::SortArrayTest_array_Test::TestBody()
                       ./fbcode/velox/functions/sparksql/tests/SortArrayTest.cpp:151
    @ 00000000001211be void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
                       fbsource/src/gtest.cc:2675
    @ 0000000000120a44 testing::Test::Run()
                       fbsource/src/gtest.cc:2692
    @ 000000000012668f testing::TestInfo::Run()
                       fbsource/src/gtest.cc:2841
    @ 000000000012e646 testing::TestSuite::Run()
                       fbsource/src/gtest.cc:3020
    @ 0000000000169fab testing::internal::UnitTestImpl::RunAllTests()
                       fbsource/src/gtest.cc:5925
    @ 000000000016900b bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
                       fbsource/src/gtest.cc:2675
    @ 0000000000168549 testing::UnitTest::Run()
                       fbsource/src/gtest.cc:5489
    @ 0000000000002980 RUN_ALL_TESTS()
                       fbsource/gtest/gtest.h:2317
                       -> ./fbcode/common/gtest/LightMain.cpp
    @ 000000000000260f main
                       ./fbcode/common/gtest/LightMain.cpp:20
    @ 000000000002c656 __libc_start_call_main
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/csu/../sysdeps/nptl/libc_start_call_main.h:58
                       -> /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/csu/../sysdeps/x86/libc-start.c
    @ 000000000002c717 __libc_start_main_alias_2
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/csu/../csu/libc-start.c:409
                       -> /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/csu/../sysdeps/x86/libc-start.c
    @ 0000000000747750 _start
                       /home/engshare/third-party2/glibc/2.34/src/glibc-2.34/csu/../sysdeps/x86_64/start.S:116

Test was never completed. The test process might have crashed.
```

boneanxs commented:
@kagamiori Hey, would you please mind sharing the relate test case? Sorry I didn't find any clue about this issue.

kagamiori commented:
> @kagamiori Hey, would you please mind sharing the relate test case? Sorry I didn't find any clue about this issue.

Hi @xiaoxmeng, could you help take a look? Thanks!

boneanxs commented:
@xiaoxmeng Hey, could you please have a look? Thanks!

xiaoxmeng commented:
> > @kagamiori Hey, would you please mind sharing the relate test case? Sorry I didn't find any clue about this issue.
> 
> Hi @xiaoxmeng, could you help take a look? Thanks!

@kagamiori I think it is the sort function issue and I run into similar issue when sort condition is changing during the sort process. Can you check if it is the case? Like inconsistent compare result between two items during the sort process

boneanxs commented:
@kagamiori @xiaoxmeng Hey, could you please trigger the fb internal test again? I fixed the strict weak ordering when the comparing values are both nulls. Thanks!

kevinwilfong commented:
It looks like there are still pending comments, so I'm removing the ready-to-merge label, feel free to add it back once they've been addressed.

kagamiori commented:
Hi @boneanxs, just a gentle reminder of this PR. It needs additional updates if you still would like to merge this change.

boneanxs commented:
> @boneanxs Would you please follow up? Thanks.

Sure, I'm addressing it
